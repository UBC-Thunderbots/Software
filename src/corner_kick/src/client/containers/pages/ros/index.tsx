import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { Main } from 'SRC/components/portals/Main';
import { Sidebar } from 'SRC/components/portals/Sidebar';
import { SidebarTitle } from 'SRC/components/portals/SidebarTitle';
import { CategoryItem } from 'SRC/components/ui/CategoryItem';
import { EmptyMain } from 'SRC/components/ui/EmptyMain';
import { RootAction } from 'SRC/store';
import { IRootState } from 'SRC/types';

import { RosIndicator } from './components/RosIndicator';

const mapStateToProps = (state: IRootState) => ({
    errorMessage: state.ros.errorMessage,
    status: state.ros.status,
});

const mapDispatchToProps = (dispatch: Dispatch<RootAction>) =>
    bindActionCreators({}, dispatch);

interface IROSPageProps {
    status: string;
    errorMessage: string;
}

class ROSPageInternal extends React.Component<IROSPageProps> {
    public render() {
        return (
            <>
                <SidebarTitle text="ROS" />
                <Sidebar>
                    <RosIndicator {...this.props} />
                    <CategoryItem text="Nodes" />
                    <CategoryItem text="Topics" />
                    <CategoryItem text="Services" />
                    <CategoryItem text="Params" />
                </Sidebar>
                <Main>
                    <EmptyMain />
                </Main>
            </>
        );
    }
}

export const ROSPage = connect(
    mapStateToProps,
    mapDispatchToProps,
)(ROSPageInternal);
