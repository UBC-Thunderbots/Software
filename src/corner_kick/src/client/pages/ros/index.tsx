import * as React from 'react';
import { connect } from 'react-redux';
import { bindActionCreators, Dispatch } from 'redux';

import { RosIndicator } from './components/RosIndicator';

import { Sidebar } from 'SRC/components/portals/Sidebar';
import { SidebarTitle } from 'SRC/components/portals/SidebarTitle';
import { CategoryItem } from 'SRC/components/ui/CategoryItem';

import { IRootState, RootAction, selectors } from 'SRC/store';

const mapStateToProps = (state: IRootState) => ({
    errorMessage: state.ros.errorMessage,
    nodes: selectors.ros.nodesSelector(state),
    params: selectors.ros.paramsSelector(state),
    services: selectors.ros.servicesSelector(state),
    status: state.ros.status,
    topics: selectors.ros.topicsSelector(state),
});

const mapDispatchToProps = (dispatch: Dispatch<RootAction>) =>
    bindActionCreators({}, dispatch);

interface IROSPageProps {
    status: string;
    errorMessage: string;
    nodes: string[];
    topics: string[];
    services: string[];
    params: string[];
}

class ROSPageInternal extends React.Component<IROSPageProps> {
    public render() {
        return (
            <>
                <SidebarTitle text="ROS" />
                <Sidebar>
                    <RosIndicator {...this.props} />
                    <CategoryItem text="Nodes" />
                    <CategoryItem text="Topics" />
                    <CategoryItem text="Services" />
                    <CategoryItem text="Params" />
                </Sidebar>
            </>
        );
    }
}

export const ROSPage = connect(
    mapStateToProps,
    mapDispatchToProps,
)(ROSPageInternal);
