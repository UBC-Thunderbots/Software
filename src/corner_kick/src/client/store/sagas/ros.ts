import { channel, delay } from 'redux-saga';
import { put, spawn, take, takeEvery, takeLatest } from 'redux-saga/effects';
import * as ROSLIB from 'roslib';

import {
    connected,
    disconnected,
    error,
    newMessage,
    setNodes,
    setParams,
    setServices,
    setTopics,
    start,
    subscribeTopic,
    unsubscribeTopic,
} from '../actions/ros';

let ros: ROSLIB.Ros | null = null;
const rosChannel = channel();

export default function* init() {
    yield takeLatest('ros/START', startROS);
    yield spawn(listenToROSChannel);
    yield put(start());
}

function* listenToROSChannel() {
    while (true) {
        const action = yield take(rosChannel);
        yield put(action);
    }
}

function* startROS() {
    yield stopROS();

    ros = new ROSLIB.Ros({});
    ros.connect('ws://localhost:9090');

    ros.on('connection', () => rosChannel.put(connected()));
    ros.on('error', () => rosChannel.put(error('There was an error')));
    ros.on('close', () => rosChannel.put(disconnected()));

    yield takeEvery('ros/SUBSCRIBE_TOPIC', subscribeToROSTopic);
    yield takeEvery('ros/UNSUBSCRIBE_TOPIC', unsubscribeFromROSTopic);

    while (true) {
        ros.getNodes((nodes) => rosChannel.put(setNodes(nodes)));
        ros.getTopics((topics) => rosChannel.put(setTopics((topics as any).topics)));
        ros.getServices((nodes) => rosChannel.put(setServices(nodes)));
        ros.getParams((nodes) => rosChannel.put(setParams(nodes)));
        yield delay(5000);
    }
}

function* stopROS() {
    if (ros !== null) {
        yield ros.close();
        ros = null;
    }
}

function subscribeToROSTopic(action: ReturnType<typeof subscribeTopic>) {
    if (ros !== null) {
        const topic = new ROSLIB.Topic({
            ros,
            messageType: action.payload.messageType,
            name: action.payload.topic,
        });

        topic.subscribe((message) =>
            rosChannel.put(newMessage(action.payload.topic, message)),
        );
    }
}

function unsubscribeFromROSTopic(action: ReturnType<typeof unsubscribeTopic>) {
    if (ros !== null) {
        const topic = new ROSLIB.Topic({
            ros,
            messageType: action.payload.messageType,
            name: action.payload.topic,
        });

        topic.unsubscribe();
    }
}
