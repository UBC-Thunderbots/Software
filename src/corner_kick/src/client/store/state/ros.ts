export interface IROSState {
    status: 'connected' | 'disconnected' | 'error';
    errorMessage: string;
    nodes: string[];
    topics: string[];
    services: string[];
    params: string[];
}
