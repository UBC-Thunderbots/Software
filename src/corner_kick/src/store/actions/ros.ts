/*
 * This file specifies ROS specific action
 *
 * We are using the format specified here
 * @see https://github.com/piotrwitek/typesafe-actions#createaction
 */

import { createAction } from 'typesafe-actions';

/**
 * Connect to ROS
 */
export const start = createAction('ros_START');

/**
 * Disconnect from ROS
 */
export const stop = createAction('ros_STOP');

/**
 * Sent when we are connected to ROS
 */
export const connected = createAction('ros_CONNECTED');

/**
 * Sent when are disconnected from ROS
 */
export const disconnected = createAction('ros_DISCONNECTED');

/**
 * Sent when there is an error connecting to ROS
 */
export const error = createAction('ros_ERROR', (resolve) => {
    return (errorMessage: string) => resolve({ errorMessage });
});

/**
 * Subscribe to a particular topic. Messages will be available through the
 * `ros/NEW_MESSAGE` action.
 */
export const subscribeTopic = createAction('ros_SUBSCRIBE_TOPIC', (resolve) => {
    return (topic: string, messageType: string) => resolve({ topic, messageType });
});

/**
 * Unsubscribes from a particular topic.
 */
export const unsubscribeTopic = createAction('ros_UNSUBSCRIBE_TOPIC', (resolve) => {
    return (topic: string, messageType: string) => resolve({ topic, messageType });
});

/**
 * Sent when a new message is received from a particular topic
 */
export const newMessage = createAction('ros_NEW_MESSAGE', (resolve) => {
    return (topic: string, message: any) => resolve({ topic, message });
});

/**
 * Sent when we receive the nodes available in ROS
 */
export const setNodes = createAction('ros_SET_NODES', (resolve) => {
    return (nodes: string[]) => resolve({ nodes });
});

/**
 * Sent when we receive the topics available in ROS
 */
export const setTopics = createAction('ros_SET_TOPICS', (resolve) => {
    return (topics: string[]) => resolve({ topics });
});

/**
 * Sent when we receive the services available in ROS
 */
export const setServices = createAction('ros_SET_SERVICES', (resolve) => {
    return (services: string[]) => resolve({ services });
});

/**
 * Sent when we receive the params available in ROS
 */
export const setParams = createAction('ros_SET_PARAMS', (resolve) => {
    return (params: string[]) => resolve({ params });
});
