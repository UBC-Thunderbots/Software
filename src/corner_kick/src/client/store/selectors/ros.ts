import { IRootState } from '../state';

export const nodesSelector = (state: IRootState) => state.ros.nodes;
export const topicsSelector = (state: IRootState) => state.ros.topics;
export const paramsSelector = (state: IRootState) => state.ros.params;
export const servicesSelector = (state: IRootState) => state.ros.services;
