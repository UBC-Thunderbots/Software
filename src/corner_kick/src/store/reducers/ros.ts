/*
 * This file specifies the ROS reducer
 */

import { ActionType, getType } from 'typesafe-actions';

import { IROSState } from 'SRC/types';

import * as ros from '../actions/ros';

export type ROSAction = ActionType<typeof ros>;

const defaultState: IROSState = {
    errorMessage: '',
    status: 'disconnected',
};

/**
 * Reducer function for ROS
 */
export default (state: IROSState = defaultState, action: ROSAction) => {
    switch (action.type) {
        // Set status to connected when we get a connected action
        case getType(ros.connected):
            return { ...state, status: 'connected' };

        // Set status to disconnected when we get a disconnected action
        case getType(ros.disconnected):
            return { ...state, status: 'disconnected' };

        // Set status to error when we get an error action
        case getType(ros.error):
            return {
                ...state,
                errorMessage: action.payload.errorMessage,
                status: 'error',
            };
        default:
            return state;
    }
};
