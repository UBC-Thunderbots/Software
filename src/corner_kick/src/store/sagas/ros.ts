/*
 * This file specifies the saga for ROS
 */

import { channel } from 'redux-saga';
import { put, spawn, take, takeLatest } from 'redux-saga/effects';
import ROSLIB from 'roslib';
import { getType } from 'typesafe-actions';

import { connected, disconnected, error, start } from '../actions/ros';

let ros: ROSLIB.Ros | null = null;
const rosChannel = channel();

/**
 * Function first called when the application first starts
 */
export default function* init() {
    // Listen to start actions and start ROS
    yield takeLatest(getType(start), startROS);

    // Start listening to ROS messages
    yield spawn(listenToROSChannel);

    // Start ROS
    yield put(start());
}

/**
 * Take any messages received from ROS and push them as Redux actions
 */
function* listenToROSChannel() {
    while (true) {
        const action = yield take(rosChannel);
        yield put(action);
    }
}

/**
 * Start ROS
 */
function* startROS() {
    yield stopROS();

    ros = new ROSLIB.Ros({});
    ros.connect('ws://localhost:9090');

    // Send Redux actions when connected, disconnected to ROS or on error
    ros.on('connection', () => rosChannel.put(connected()));
    ros.on('error', () => rosChannel.put(error('There was an error')));
    ros.on('close', () => rosChannel.put(disconnected()));
}

/**
 * Disconnect from ROS if we are connected
 */
function* stopROS() {
    if (ros !== null) {
        yield ros.close();
        ros = null;
    }
}

/**
 * Subscribe to a ROS topic and send messages to the callback
 */
export function subscribeToROSTopic(
    name: string,
    messageType: string,
    callback: (message: ROSLIB.Message) => void,
) {
    if (ros !== null) {
        const topic = new ROSLIB.Topic({
            messageType,
            name,
            ros,
        });

        topic.subscribe(callback);
    }
}

/**
 * Unsubscribe from a ROS topic
 */
export function unsubscribeToROSTopic(
    name: string,
    messageType: string,
    callback: (message: ROSLIB.Message) => void,
) {
    if (ros !== null) {
        const topic = new ROSLIB.Topic({
            messageType,
            name,
            ros,
        });

        topic.unsubscribe(callback);
    }
}
