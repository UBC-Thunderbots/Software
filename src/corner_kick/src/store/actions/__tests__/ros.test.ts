/**
 * This file is for testing ROS actions
 * Each test case has a description of what it tests
 */

import * as rosActions from '../ros';

describe('rosActions', () => {
    describe('when we connect to ROS', () => {
        it('dispatches the correct start action and payload', () => {
            expect(rosActions.start()).toMatchSnapshot();
        });
    });

    describe('when we disconnect to ROS', () => {
        it('dispatches the correct stop action and payload', () => {
            expect(rosActions.stop()).toMatchSnapshot();
        });
    });

    describe('when we are connected to ROS', () => {
        it('dispatches the correct connect action and payload', () => {
            expect(rosActions.connected()).toMatchSnapshot();
        });
    });

    describe('when we are disconnected to ROS', () => {
        it('dispatches the correct disconnect action and payload', () => {
            expect(rosActions.disconnected()).toMatchSnapshot();
        });
    });

    describe('when there is an error connecting to ROS', () => {
        it('dispatches the correct error action and payload', () => {
            expect(rosActions.error('Error getting pan-tilt limit')).toMatchSnapshot();
        });
    });
});
