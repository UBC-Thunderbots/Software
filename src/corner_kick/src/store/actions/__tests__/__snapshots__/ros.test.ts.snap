// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rosActions when there is an error connecting to ROS dispatches the correct error action and payload 1`] = `
Object {
  "meta": undefined,
  "payload": Object {
    "errorMessage": "Error getting pan-tilt limit",
  },
  "type": "ros_ERROR",
}
`;

exports[`rosActions when we are connected to ROS dispatches the correct connect action and payload 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "ros_CONNECTED",
}
`;

exports[`rosActions when we are disconnected to ROS dispatches the correct disconnect action and payload 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "ros_DISCONNECTED",
}
`;

exports[`rosActions when we connect to ROS dispatches the correct start action and payload 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "ros_START",
}
`;

exports[`rosActions when we disconnect to ROS dispatches the correct stop action and payload 1`] = `
Object {
  "meta": undefined,
  "payload": undefined,
  "type": "ros_STOP",
}
`;
